client/components/tinymce/plugins/mentions/mentions.jsx

```diff
@@ -0,0 +1,218 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import ReactDomServer from 'react-dom/server';
+import { connect } from 'react-redux';
+import { includes } from 'lodash';
+import tinymce from 'tinymce/tinymce';
+
+/**
+ * Internal dependencies
+ */
+import SuggestionList from './suggestion-list';
+import EditorMention from './editor-mention';
+import QueryUsersSuggestions from 'components/data/query-users-suggestions';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getUserSuggestions } from 'state/users/suggestions/selectors';
+
+/**
+ * Module variables
+ */
+const VK = tinymce.util.VK;
+
+const getMatchingSuggestions = function( suggestions, query ) {
+	let matchingSuggestions = suggestions;
+
+	if ( query ) {
+		// Start of string or preceded by a space.
+		const matcher = new RegExp( '^' + query + '| ' + query, 'ig' );
+
+		matchingSuggestions = suggestions.filter( ( suggestion ) => {
+			const name = suggestion.name || suggestion.user_login + ' ' + suggestion.display_name;
+
+			return matcher.test( name );
+		} );
+	}
+
+	return matchingSuggestions.slice( 0, 10 );
+};
+
+export class Mentions extends React.Component {
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			query: '',
+			popoverContext: null,
+			showPopover: false,
+		};
+	}
+
+	componentDidMount() {
+		const { editor } = this.props;
+		const position = this.getPosition();
+
+		editor.on( 'keyup', this.onKeyUp );
+		editor.on( 'click', this.hidePopover );
+
+		this.left = position.left;
+		this.top = position.top;
+	}
+
+	componentWillUpdate( nextProps, nextState ) {
+		// Update position of popover if going from invisible to visible state.
+		if ( ! this.state.showPopover && nextState.showPopover ) {
+			this.updatePosition();
+			return;
+		}
+
+		// Update position of popover if cursor has moved to a new line.
+		if ( nextState.showPopover ) {
+			const currentPosition = this.getPosition();
+			const isLineBefore = ( this.top > currentPosition.top ) && ( this.left < currentPosition.left );
+			const isLineAfter = ( this.top < currentPosition.top ) && ( this.left > currentPosition.left );
+
+			if ( isLineBefore || isLineAfter ) {
+				this.updatePosition( currentPosition );
+			}
+		}
+	}
+
+	componentWillUnmount() {
+		const { editor } = this.props;
+
+		editor.off( 'keyup', this.onKeyUp );
+		editor.off( 'click', this.hidePopover );
+	}
+
+	getPosition() {
+		let left, top;
+		const { editor } = this.props;
+		const nodePosition = tinymce.$( editor.selection.getNode() ).offset();
+		const rectList = editor.selection.getRng().getClientRects();
+
+		if ( rectList.length > 0 ) {
+			left = rectList[ 0 ].left;
+			top = rectList[ 0 ].top + rectList[ 0 ].height;
+		} else {
+			left = nodePosition.left;
+			top = nodePosition.top;
+		}
+
+		return { left: left, top: top };
```

Comment:
do we need to set these variables and then return later? maybe it would make sense to return early here too

```js
if ( rectList.length > 0 ) {
	return {
		left: rectList[ 0 ].left,
		top: rectList[ 0 ].top + rectList[ 0 ].height,
	};
}

return {
	left: nodePosition.left,
	top: nodePosition.top,
};
```

could have also used a helper here from `lodash` when simply plucking out variables but not necessary by any means

```js
return pick( nodePosition, [ 'left', 'top' ] );
```